src/ios-scroll.js

```diff
@@ -0,0 +1,557 @@
+ * limitations under the License.
+ */
+
+
+/**
+ * The elapsed time cutoff that determines if a user intended to stop the
+ * synthetic scroll momentum.
+ * @const @private {number}
+ */
+const SCROLL_INTENTION_ELAPSED_CUTOFF = (1000 / 60) * 3;
+
+
+/**
+ * The number of pixels necessary to fire a scroll event. A scroll event will
+ * always fire at the end of a scroll, but only after this many pixels have
+ * changed during the scroll.
+ * @const @private {number}
+ */
+const SCROLL_EVENT_DELTA = 2.1;
+
+/**
+ * Manages scrolling of embedded AMP document in iOS, since iOS doesn't
+ * natively scroll embedded documents well. This "synthetic" scrolling keeps
+ * track of fixed position elements so that they may be scrolled in unison. It
+ * also keeps track of iframes so that they can be disabled during scrolling
+ * (so they don't start a native scroll during our synthetic scroll).
+ *
+ * This is a substantially modified version of Kinetic Scrolling.
+ */
+export class SyntheticScroll {
+
+  /**
+   * @param {HTMLBodyElement} body
+   * @param {number} scrollTop the body's current scrollTop
+   * @param {number} scrollHeight the body's current scrollHeight
+   */
+  constructor(scrollEl, scrollTop, scrollHeight) {
+    const doc = scrollEl.ownerDocument;
+    const body = doc.body;
+
+    /**
+     * @const @private {!Window}
+     */
+    this.win_ = doc.defaultView;
+
+    /**
+     * We store the style reference directly to speed things up.
+     * @const @private {!CSSStyleDeclaration}
+     **/
+    this.scrollStyle_ = scrollEl.style;
+
+    /**
+     * We track the iframes on the page so we may disable direct interaction
+     * during scrolling. This prevents them from causing native scrolling.
+     * When scrolling is done, they will be re-enabled.
+     * @const @private {!Array<!CSSStyleDeclaration>}
+     **/
+    this.iframeStyles_ = [];
+
+    /**
+     * A list of fixed position elements that we should scroll with the doc.
+     * @const @private {!Array<!CSSStyleDeclaration>}
+     */
+    this.fixedElementStyles_ = [];
+
+    /**
+     * A helper element that records the last position we scrolled to. This
+     * allows us to determine how much the user scrolled natively.
+     * @const @private {Element}
+     */
+    this.scrollMoveEl_ = doc.createElement('div');
+    this.scrollMoveEl_.id = '-amp-scrollmove';
+    body.appendChild(this.scrollMoveEl_);
+
+    /**
+     * The body's current scrollTop position. This is synchronized after every
+     * synthetic scroll and at the start of the synthetic scroll after a native
+     * scroll.
+     * @private {number}
+     */
+    this.scrollTop_ = scrollTop;
+
+    /**
+     * The minimum value that we may scroll the scroller to. This may be less
+     * than zero so we may "overscroll" upwards after synchronizing the body's
+     * scrollTop and our synthetic scroll offset.
+     * @private {number}
+     */
+    this.min_ = -scrollTop;
+
```

Comment:
can we rename this and max to minScrollTop/maxScrollTop, it gets confusing when i read it in the code way below.